src/modules.ml

```diff
@@ -134,6 +134,85 @@ module Stdlib = struct
       lib_interface t
 end
 
+module Mangle = struct
+  module Lib = struct
+    type kind =
+      | Has_lib_interface
+      | Implementation of Lib_name.Local.t
+      | Neither
+
+    type t =
+      { main_module_name : Module.Name.t
+      ; kind : kind
+      }
+  end
+
+  type t =
+    | Lib of Lib.t
+    | User_exe of Path.Build.t
+    | Internal of string * Lib_name.Local.t option
+
+  let internal_exe ~purpose ~for_ = Internal (purpose, for_)
+
+  let of_lib ~main_module_name ~modules ~(lib : Dune_file.Library.t) =
+    let kind : Lib.kind =
+      if Option.is_some lib.implements then
+        Implementation (snd lib.name)
+      else if Module.Name.Map.mem modules main_module_name then
+        Has_lib_interface
+      else
+        Neither
+    in
+    Lib { main_module_name ; kind }
+
+  let prefix t : Module.Name.t Visibility.Map.t =
+    match t with
+    | Lib { main_module_name ; kind } ->
+      begin match kind with
+      | Has_lib_interface
+      | Neither -> Visibility.Map.make_both main_module_name
+      | Implementation lib ->
+        { private_ =
+            sprintf "%s__%s__"
+              (Module.Name.to_string main_module_name)
+              (Lib_name.Local.to_string lib)
+            |> Module.Name.of_string
+        ; public = main_module_name
+        }
+      end
+    | Internal (purpose, for_) ->
```

Comment:
We can go even further than removing the project name in the prefix. We can get rid of the Internal/User distinction and systematically use `Dune__exe__` as prefix. That'll make things simpler.

Reply:
Okay. Shall we keep `purpose` and `for_` around however?

Reply:
Nope, if there are not used for anything meaningful, let's get rid of them. Otherwise it's more data to maintain.